proposals/intraprocess_communication.md

```diff
@@ -0,0 +1,525 @@
+scenario it will not be able to find anything and this will only result in a warning message. However, if some
+`Publisher` with intra-process communication enabled happens to be present also in that process, then the
+`Subscription` will end up extracting something from the ring buffer. This is due to the fact that the id of
+the publisher contained in the meta-message is just an integer value guaranteed to be unique only within the
+process that generated it.
+
+This situation has two possible outcomes: if the `Publisher` from which the message is extracted has the same
+message type as the `Subscription`, then the problem will not be noticed, even if the `Subscription` is taking
+a message that can potentially be old or destinated to a different consumer. On the other hand, if the
+`Publisher` from which the message is extracted has a different message type from the one of the
+`Subscription`, the `static_cast` operations performed by the `IntraProcessManager` will cause an access
+violation error.
+
+
+## Proposed implementation
+
+### Overview
+
+The new proposal for intra-process communication addresses all the issues previously mentioned. It has been
+designed with performance in mind, so it avoids any communication through the middleware between nodes in the
+same process. Moreover, it supports all the ROS2 Quality of Service configurations.
+
+Consider a simple scenario, consisting of `Publisher`s and `Subscription`s all in the same process and with
+the durability QoS set to `volatile`. The proposed implementation creates one buffer per `Subscription`. When
+a message is published to a topic, its `Publisher` pushes the message into the buffer of each of the
+`Subscription`s related to that topic and raises a notification, waking up the executor. The executor can then pop
+the message from the buffer and trigger the callback of the `Subscription`.
+
+![Proposed IPC Block Diagram](new_ipc.png)
+
+There are three possible data-types that can be stored in the buffer:
+
+- `MessageT`
+- `shared_ptr<const MessageT>`
+- `unique_ptr<MessageT>`
+
+The choice is left to the user. By default a `Subscription` will use the type between `shared_ptr<constMessageT>`
+and `unique_ptr<MessageT>` that better fits with its callback type.
+
+If the history QoS is set to `keep all`, the buffers are dynamically allocated. On the other hand, if the
+history QoS is set to `keep last`, the buffers have a size equal to the depth of the history and they act as
+ring buffers (overwriting the oldest data when trying to push while its full). Ring buffers are not only used
+in `Subscription`s but also in each `Publisher` with a durability QoS of type `transient local`.
+
+The `IntraProcessManager` class has access to all the ring buffers and has also information about which `Publisher`s and
+`Subscription`s are connected to each other.
+
+A new class derived from `rclcpp::Waitable` is defined, denominated `IPCWaitable`. An object of this type is
```

Comment:
As @wjwwood mentioned, use of `IPC` is ambiguous.  Probably change this to `IntraWaitable` or similar.

Reply:
Oh yes, I updated `IPC` and I forgot about this one..